src/stocastic_functions/propose.jl

```diff
@@ -0,0 +1,73 @@
+# Arguments
+- `current_program::RuleNode`: the current program.
+- `neighbourhood_node_loc::NodeLoc`: the location of the program to replace.
+- `grammar::Grammar`: the grammar used to create programs.
+- `max_depth::Int`: the maximum depth of the resulting programs.
+- `dict::Dict{String, Any}`: the dictionary with additional arguments; not used.
+"""
+function random_fill_propose(current_program, neighbourhood_node_loc, grammar, max_depth, dict)
+    # it can change the current_program for fast replacing of the node
+    # find the symbol of subprogram
+    subprogram = get(current_program, neighbourhood_node_loc)
+    neighbourhood_symbol = return_type(grammar, subprogram)
+
+    # find the depth of subprogram 
+    current_depth = node_depth(current_program, subprogram)
+    # this is depth that we can still generate without exceeding max_depth
+    remaining_depth = max_depth - current_depth + 1
+
+    if remaining_depth == 0
+        # can't expand more => return current program 
+        @warn "Can't extend program because we reach max_depth $(rulenode2expr(current_program, grammar))"
+        return [current_program]
+    end
+
+    @assert remaining_depth >= 1 "remaining_depth $remaining_depth should be bigger than 1 here"
+    # generate completely random expression (subprogram) with remaining_depth
+    replacement = rand(RuleNode, grammar, neighbourhood_symbol, remaining_depth)
+    @assert depth(replacement) <= remaining_depth "The depth of new random = $(depth(replacement)) but remaning depth =  $remaining_depth. 
+            Expreesion was $(rulenode2expr(current_program,grammar))"
+
+    @assert depth(current_program) <= max_depth "Depth of program is $(depth(current_program)) but max_depth = $max_depth"
+    return [replacement]
+end
+
+"""
+The return function is a function that produces a list with all the subprograms constructed by using a subset of the grammar rules with depth at most `enumeration_depth`.
+The function expects the entry with key "rule_subset" in `dict` and value of type Vector{Any}.
+# Arguments
+- `current_program::Int64`: the maximum enumeration depth.
+"""
+function enumerate_neighbours_propose(enumeration_depth)
+    return (current_program, neighbourhood_node_loc, grammar, max_depth, dict) -> begin
+        # it can change the current_program for fast replacing of the node
+        # find the symbol of subprogram
+        subprogram = get(current_program, neighbourhood_node_loc)
+        neighbourhood_symbol = return_type(grammar, subprogram)
+    
+        # find the depth of subprogram
+        current_depth = node_depth(current_program, subprogram)
+        # this is depth that we can still generate without exceeding max_depth
+        remaining_depth = max_depth - current_depth + 1  
+        depth_left = min(remaining_depth, enumeration_depth)
+        subset_grammar = ContextFreeGrammar(dict["rule_subset"], grammar.types, grammar.isterminal,
```

Comment:
change `contextFreeGrammar` to `ContextSensitiveGrammar`

Reply:
we changed.